File: tests/python/test_tupledns.py
"""
TupleDNS Python Test Suite

Comprehensive tests for the Python bindings using pytest
"""

import pytest
import sys
import os
import time
from typing import List, Optional

# Add parent directory to path to import tupledns
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))
import tupledns

class TestTupleDNSCore:
    """Test core TupleDNS functionality"""
    
    def setup_method(self):
        """Setup for each test method"""
        tupledns.init()
    
    def teardown_method(self):
        """Cleanup after each test method"""
        tupledns.cleanup()
    
    def test_library_lifecycle(self):
        """Test library initialization and cleanup"""
        tupledns.cleanup()
        tupledns.init()  # Should not raise
        tupledns.cleanup()
        tupledns.init()  # Should work multiple times
    
    def test_coordinate_validation(self):
        """Test coordinate validation"""
        # Valid coordinates
        assert tupledns.validate_coordinate("simple.test.tuple")
        assert tupledns.validate_coordinate("a.b.c.d.e.f.tuple")
        assert tupledns.validate_coordinate("123.456.789.numeric.tuple")
        assert tupledns.validate_coordinate("with-hyphens.and-more.test.tuple")
        
        # Invalid coordinates
        assert not tupledns.validate_coordinate("")
        assert not tupledns.validate_coordinate("no-suffix")
        assert not tupledns.validate_coordinate(".tuple")
        assert not tupledns.validate_coordinate("spaces not allowed.tuple")
        assert not tupledns.validate_coordinate("under_scores.tuple")
        assert not tupledns.validate_coordinate("special@chars.tuple")
    
    def test_coordinate_encoding(self):
        """Test coordinate encoding and decoding"""
        components = ["ambient", "120", "london", "music"]
        encoded = tupledns.encode_coordinate(components)
        assert encoded == "ambient.120.london.music.tuple"
        
        decoded = tupledns.decode_coordinate(encoded)
        assert decoded == components
    
    def test_pattern_matching(self):
        """Test pattern matching functionality"""
        coord = "ambient.120.london.music.tuple"
        
        # Exact matches
        assert tupledns.match_pattern(coord, coord)
        
        # Wildcard matches
        assert tupledns.match_pattern(coord, "*.120.london.music.tuple")
        assert tupledns.match_pattern(coord, "ambient.*.london.music.tuple")
        assert tupledns.match_pattern(coord, "ambient.120.*.music.tuple")
        assert tupledns.match_pattern(coord, "*.*.*.*.tuple")
        
        # Non-matches
        assert not tupledns.match_pattern(coord, "jazz.120.london.music.tuple")
        assert not tupledns.match_pattern(coord, "ambient.140.london.music.tuple")
        assert not tupledns.match_pattern(coord, "ambient.120.london.music.spatial.tuple")

class TestTupleDNSRegistration:
    """Test registration and discovery functionality"""
    
    def setup_method(self):
        tupledns.init()
    
    def teardown_method(self):
        tupledns.cleanup()
    
    def create_test_coordinate(self, prefix: str) -> str:
        """Create a unique test coordinate"""
        timestamp = int(time.time() * 1000) % 10000
        return f"{prefix}.test.{timestamp}.tuple"
    
    def test_basic_registration(self):
        """Test basic node registration"""
        coord = self.create_test_coordinate("basic")
        capabilities = ["test", "python"]
        
        # Should not raise
        tupledns.register(coord, capabilities, ttl=300)
        tupledns.unregister(coord)
    
    def test_registration_with_capabilities(self):
        """Test registration with various capability combinations"""
        coords_and_caps = [
            (self.create_test_coordinate("music"), ["midi", "real-time", "collaborative"]),
            (self.create_test_coordinate("iot"), ["sensor", "temperature", "wireless"]),
            (self.create_test_coordinate("ai"), ["reasoning", "creativity", "multimodal"]),
        ]
        
        for coord, caps in coords_and_caps:
            tupledns.register(coord, caps)
        
        # Cleanup
        for coord, _ in coords_and_caps:
            tupledns.unregister(coord)
    
    def test_discovery_basic(self):
        """Test basic discovery functionality"""
        # Register some test nodes
        coord1 = self.create_test_coordinate("discover1")
        coord2 = self.create_test_coordinate("discover2")
        
        tupledns.register(coord1, ["test", "discovery"])
        tupledns.register(coord2, ["test", "discovery"])
        
        # Discover nodes
        result = tupledns.find("*.test.*.tuple")
        assert result is not None
        assert hasattr(result, 'nodes')
        assert hasattr(result, 'total_queries')
        assert hasattr(result, 'query_time')
        
        # Cleanup
        tupledns.unregister(coord1)
        tupledns.unregister(coord2)
    
    def test_capability_filtering(self):
        """Test discovery with capability filtering"""
        coord1 = self.create_test_coordinate("cap1")
        coord2 = self.create_test_coordinate("cap2")
        
        tupledns.register(coord1, ["common", "unique1"])
        tupledns.register(coord2, ["common", "unique2"])
        
        # Find with capability filter
        result = tupledns.find("*.test.*.tuple", capabilities=["common"])
        assert result is not None
        
        result_unique = tupledns.find("*.test.*.tuple", capabilities=["unique1"])
        assert result_unique is not None
        
        # Cleanup
        tupledns.unregister(coord1)
        tupledns.unregister(coord2)

class TestTupleDNSRangeQueries:
    """Test range query functionality"""
    
    def setup_method(self):
        tupledns.init()
    
    def teardown_method(self):
        tupledns.cleanup()
    
    def test_numeric_range(self):
        """Test numeric range queries"""
        coords = [
            "ambient.100.test.music.tuple",
            "ambient.120.test.music.tuple", 
            "ambient.140.test.music.tuple"
        ]
        
        for coord in coords:
            tupledns.register(coord, ["music", "test"])
        
        # Range query
        result = tupledns.find_range(
            "ambient.{bpm}.test.music.tuple",
            ranges={"bpm": (110, 130)}
        )
        assert result is not None
        
        # Cleanup
        for coord in coords:
            tupledns.unregister(coord)
    
    def test_multi_dimensional_range(self):
        """Test multi-dimensional range queries"""
        coords = [
            "ambient.120.london.music.tuple",
            "jazz.120.newyork.music.tuple",
            "electronic.140.berlin.music.tuple"
        ]
        
        for coord in coords:
            tupledns.register(coord, ["music"])
        
        # Multi-dimensional range
        result = tupledns.find_range(
            "{genre}.{bpm}.{city}.music.tuple",
            ranges={
                "genre": ["ambient", "jazz"],
                "bpm": (110, 130),
                "city": ["london", "newyork", "berlin"]
            }
        )
        assert result is not None
        
        # Cleanup
        for coord in coords:
            tupledns.unregister(coord)

class TestTupleDNSErrorHandling:
    """Test error handling and edge cases"""
    
    def test_invalid_coordinates(self):
        """Test handling of invalid coordinates"""
        tupledns.init()
        
        with pytest.raises(Exception):
            tupledns.register("", ["test"])
        
        with pytest.raises(Exception):
            tupledns.register("invalid coordinate", ["test"])
        
        with pytest.raises(Exception):
            tupledns.find("")
        
        tupledns.cleanup()
    
    def test_none_arguments(self):
        """Test handling of None arguments"""
        tupledns.init()
        
        with pytest.raises(Exception):
            tupledns.register(None, ["test"])
        
        with pytest.raises(Exception):
            tupledns.find(None)
        
        tupledns.cleanup()
    
    def test_empty_capabilities(self):
        """Test registration with empty capabilities"""
        tupledns.init()
        
        coord = f"empty.{int(time.time())}.test.tuple"
        # Should work with empty capabilities
        tupledns.register(coord, [])
        tupledns.unregister(coord)
        
        tupledns.cleanup()

class TestTupleDNSPerformance:
    """Performance and stress tests"""
    
    def setup_method(self):
        tupledns.init()
    
    def teardown_method(self):
        tupledns.cleanup()
    
    @pytest.mark.performance
    def test_bulk_registration(self):
        """Test bulk registration performance"""
        coords = []
        start_time = time.time()
        
        # Register 100 nodes
        for i in range(100):
            coord = f"perf.{i}.bulk.test.tuple"
            coords.append(coord)
            tupledns.register(coord, ["performance", "test"])
        
        reg_time = time.time() - start_time
        print(f"\nBulk registration: 100 nodes in {reg_time:.3f}s")
        
        # Cleanup
        for coord in coords:
            tupledns.unregister(coord)
    
    @pytest.mark.performance
    def test_discovery_performance(self):
        """Test discovery performance"""
        # Register test nodes
        coords = []
        for i in range(50):
            coord = f"discover.{i}.perf.test.tuple"
            coords.append(coord)
            tupledns.register(coord, ["performance"])
        
        start_time = time.time()
        
        # Perform multiple discoveries
        for _ in range(20):
            result = tupledns.find("discover.*.perf.test.tuple")
        
        disc_time = time.time() - start_time
        print(f"\nDiscovery performance: 20 queries in {disc_time:.3f}s")
        
        # Cleanup
        for coord in coords:
            tupledns.unregister(coord)
    
    @pytest.mark.stress
    def test_memory_stress(self):
        """Test memory usage under stress"""
        coords = []
        
        # Register many nodes with varying capabilities
        for i in range(200):
            coord = f"stress.{i}.memory.test.tuple"
            coords.append(coord)
            caps = [f"cap{j}" for j in range(i % 5 + 1)]
            tupledns.register(coord, caps)
        
        # Perform many discoveries
        for _ in range(50):
            result = tupledns.find("stress.*.memory.test.tuple")
            if result:
                # Force some processing
                len(result.nodes)
        
        # Cleanup
        for coord in coords:
            tupledns.unregister(coord)

class TestTupleDNSRealWorldScenarios:
    """Test real-world usage scenarios"""
    
    def setup_method(self):
        tupledns.init()
    
    def teardown_method(self):
        tupledns.cleanup()
    
    def test_music_collaboration_scenario(self):
        """Test music collaboration use case"""
        # Register musicians
        musicians = [
            ("ambient.120.experimental.london.music.tuple", ["midi", "real-time", "collaborative"]),
            ("jazz.140.swing.newyork.music.tuple", ["midi", "live-recording", "improvisation"]),
            ("electronic.128.minimal.berlin.music.tuple", ["midi", "synthesis", "real-time"]),
        ]
        
        for coord, caps in musicians:
            tupledns.register(coord, caps)
        
        # Find collaborators by BPM range
        result = tupledns.find_range(
            "{genre}.{bpm}.{style}.{city}.music.tuple",
            ranges={"bpm": (115, 135)},
            capabilities=["real-time"]
        )
        assert result is not None
        
        # Find MIDI-capable musicians
        midi_result = tupledns.find("*.*.*.*.music.tuple", capabilities=["midi"])
        assert midi_result is not None
        
        # Cleanup
        for coord, _ in musicians:
            tupledns.unregister(coord)
    
    def test_iot_device_scenario(self):
        """Test IoT device management use case"""
        # Register IoT devices
        devices = [
            ("sensor-1.kitchen.floor-1.building-5.spatial.tuple", ["temperature", "humidity", "iot"]),
            ("printer-1.lab.floor-2.building-5.spatial.tuple", ["3d-print", "pla", "petg"]),
            ("light-1.office-201.floor-2.building-5.spatial.tuple", ["lighting", "dimmer", "color"]),
        ]
        
        for coord, caps in devices:
            tupledns.register(coord, caps)
        
        # Find all devices in building
        building_devices = tupledns.find("*.*.*.building-5.spatial.tuple")
        assert building_devices is not None
        
        # Find sensors specifically
        sensors = tupledns.find("*.*.*.building-5.spatial.tuple", capabilities=["iot"])
        assert sensors is not None
        
        # Find devices on specific floor
        floor2_devices = tupledns.find("*.floor-2.building-5.spatial.tuple")
        assert floor2_devices is not None
        
        # Cleanup
        for coord, _ in devices:
            tupledns.unregister(coord)
    
    def test_ai_agent_scenario(self):
        """Test AI agent discovery use case"""
        # Register AI agents
        agents = [
            ("creative.80.collaborative.research.ai.tuple", ["reasoning", "creativity", "multimodal"]),
            ("analytical.120.autonomous.production.ai.tuple", ["reasoning", "analysis", "code-generation"]),
            ("empathetic.70.collaborative.support.ai.tuple", ["reasoning", "empathy", "writing"]),
        ]
        
        for coord, caps in agents:
            tupledns.register(coord, caps)
        
        # Find collaborative agents
        collaborative = tupledns.find("*.*.collaborative.*.ai.tuple")
        assert collaborative is not None
        
        # Find agents with reasoning capability
        reasoning_agents = tupledns.find("*.*.*.*.ai.tuple", capabilities=["reasoning"])
        assert reasoning_agents is not None
        
        # Find agents by frequency range
        freq_range = tupledns.find_range(
            "{personality}.{freq}.{mode}.{context}.ai.tuple",
            ranges={"freq": (60, 90)}
        )
        assert freq_range is not None
        
        # Cleanup
        for coord, _ in agents:
            tupledns.unregister(coord)

# Test configuration for pytest
if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/javascript/test_tupledns.js
/**
 * TupleDNS JavaScript Test Suite
 * 
 * Comprehensive tests for JavaScript/Node.js bindings
 */

const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');

// Load TupleDNS JavaScript bindings
const TupleDNS = require('../../tupledns.js');

class TestSuite {
    constructor() {
        this.tests = [];
        this.passed = 0;
        this.failed = 0;
        this.proxyProcess = null;
        this.proxyPort = 3003;
    }

    addTest(name, testFunc) {
        this.tests.push({ name, testFunc });
    }

    async runTests() {
        console.log('TupleDNS JavaScript Test Suite');
        console.log('==============================\n');

        // Start proxy server for browser tests
        await this.startProxy();

        for (const test of this.tests) {
            try {
                console.log(`Running: ${test.name}`);
                await test.testFunc();
                console.log('  ✅ PASS\n');
                this.passed++;
            } catch (error) {
                console.log(`  ❌ FAIL: ${error.message}\n`);
                this.failed++;
            }
        }

        await this.stopProxy();

        console.log('==============================');
        console.log(`Results: ${this.passed} passed, ${this.failed} failed`);
        console.log(`Status: ${this.failed === 0 ? 'ALL TESTS PASSED! ✅' : 'SOME TESTS FAILED ❌'}`);

        return this.failed === 0;
    }

    async startProxy() {
        return new Promise((resolve, reject) => {
            this.proxyProcess = spawn('node', [
                path.join(__dirname, '../../proxy-server.cjs'),
                '--port', this.proxyPort.toString()
            ], {
                stdio: ['pipe', 'pipe', 'pipe']
            });

            this.proxyProcess.stdout.on('data', (data) => {
                if (data.toString().includes('Ready to proxy TupleDNS requests')) {
                    setTimeout(resolve, 500);
                }
            });

            this.proxyProcess.stderr.on('data', (data) => {
                console.error('Proxy error:', data.toString());
            });

            setTimeout(resolve, 3000); // Fallback timeout
        });
    }

    async stopProxy() {
        if (this.proxyProcess) {
            this.proxyProcess.kill();
            await new Promise(resolve => {
                this.proxyProcess.on('exit', resolve);
                setTimeout(resolve, 1000);
            });
        }
    }

    assert(condition, message) {
        if (!condition) {
            throw new Error(message);
        }
    }

    assertEqual(actual, expected, message) {
        if (actual !== expected) {
            throw new Error(`${message}: expected ${expected}, got ${actual}`);
        }
    }

    assertArrayEqual(actual, expected, message) {
        if (JSON.stringify(actual) !== JSON.stringify(expected)) {
            throw new Error(`${message}: expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
        }
    }
}

// Test suite instance
const suite = new TestSuite();

// Core functionality tests
suite.addTest('TupleDNS Class Creation', async () => {
    const dns = new TupleDNS.TupleDNS();
    suite.assert(dns instanceof TupleDNS.TupleDNS, 'Should create TupleDNS instance');
});

suite.addTest('TupleDNS Initialization', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    suite.assert(true, 'Initialization should complete without error');
});

suite.addTest('Coordinate Validation', async () => {
    // Valid coordinates
    suite.assert(TupleDNS.validateCoordinate('simple.test.tuple'), 'Simple coordinate should be valid');
    suite.assert(TupleDNS.validateCoordinate('a.b.c.d.e.f.tuple'), 'Long coordinate should be valid');
    suite.assert(TupleDNS.validateCoordinate('123.456.numeric.tuple'), 'Numeric values should be valid');
    suite.assert(TupleDNS.validateCoordinate('with-hyphens.test.tuple'), 'Hyphens should be valid');

    // Invalid coordinates
    suite.assert(!TupleDNS.validateCoordinate(''), 'Empty string should be invalid');
    suite.assert(!TupleDNS.validateCoordinate('no-suffix'), 'Missing .tuple should be invalid');
    suite.assert(!TupleDNS.validateCoordinate('.tuple'), 'Just .tuple should be invalid');
    suite.assert(!TupleDNS.validateCoordinate('spaces not allowed.tuple'), 'Spaces should be invalid');
});

suite.addTest('Coordinate Encoding', async () => {
    const components = ['ambient', '120', 'london', 'music'];
    const encoded = TupleDNS.encodeCoordinate(components);
    suite.assertEqual(encoded, 'ambient.120.london.music.tuple', 'Encoded coordinate should match expected');
    
    const decoded = TupleDNS.decodeCoordinate(encoded);
    suite.assertArrayEqual(decoded, components, 'Decoded components should match original');
});

suite.addTest('Pattern Matching', async () => {
    const coord = 'ambient.120.london.music.tuple';
    
    // Exact matches
    suite.assert(TupleDNS.matchPattern(coord, coord), 'Exact match should work');
    
    // Wildcard matches
    suite.assert(TupleDNS.matchPattern(coord, '*.120.london.music.tuple'), 'First wildcard should match');
    suite.assert(TupleDNS.matchPattern(coord, 'ambient.*.london.music.tuple'), 'Middle wildcard should match');
    suite.assert(TupleDNS.matchPattern(coord, 'ambient.120.*.music.tuple'), 'Last wildcard should match');
    suite.assert(TupleDNS.matchPattern(coord, '*.*.*.*.tuple'), 'All wildcards should match');
    
    // Non-matches
    suite.assert(!TupleDNS.matchPattern(coord, 'jazz.120.london.music.tuple'), 'Different component should not match');
    suite.assert(!TupleDNS.matchPattern(coord, 'ambient.140.london.music.tuple'), 'Different component should not match');
});

suite.addTest('Node Registration', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    
    const coordinate = `test.${Date.now()}.registration.tuple`;
    const capabilities = ['javascript', 'test', 'registration'];
    
    // Should not throw
    await dns.register(coordinate, capabilities, 300);
    await dns.unregister(coordinate);
});

suite.addTest('Node Discovery', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    
    // Register test nodes
    const coord1 = `discover1.${Date.now()}.test.tuple`;
    const coord2 = `discover2.${Date.now()}.test.tuple`;
    
    await dns.register(coord1, ['test', 'discovery']);
    await dns.register(coord2, ['test', 'discovery']);
    
    // Discover nodes
    const result = await dns.find('*.*.test.tuple');
    suite.assert(result instanceof TupleDNS.TupleResult, 'Should return TupleResult instance');
    suite.assert(Array.isArray(result.nodes), 'Should have nodes array');
    suite.assert(typeof result.totalQueries === 'number', 'Should have totalQueries');
    suite.assert(typeof result.queryTime === 'number', 'Should have queryTime');
    
    // Cleanup
    await dns.unregister(coord1);
    await dns.unregister(coord2);
});

suite.addTest('Capability Filtering', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    
    const coord1 = `cap1.${Date.now()}.test.tuple`;
    const coord2 = `cap2.${Date.now()}.test.tuple`;
    
    await dns.register(coord1, ['common', 'unique1']);
    await dns.register(coord2, ['common', 'unique2']);
    
    // Find with capability filter
    const commonResult = await dns.find('*.*.test.tuple', ['common']);
    suite.assert(commonResult instanceof TupleDNS.TupleResult, 'Should return results for common capability');
    
    const uniqueResult = await dns.find('*.*.test.tuple', ['unique1']);
    suite.assert(uniqueResult instanceof TupleDNS.TupleResult, 'Should return results for unique capability');
    
    // Cleanup
    await dns.unregister(coord1);
    await dns.unregister(coord2);
});

suite.addTest('Range Queries', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    
    const coords = [
        `ambient.100.${Date.now()}.music.tuple`,
        `ambient.120.${Date.now()}.music.tuple`,
        `ambient.140.${Date.now()}.music.tuple`
    ];
    
    for (const coord of coords) {
        await dns.register(coord, ['music', 'test']);
    }
    
    // Range query
    const result = await dns.findRange(
        'ambient.{bpm}.{timestamp}.music.tuple',
        { 
            bpm: [100, 120, 140],
            timestamp: [Date.now()]
        }
    );
    
    suite.assert(result instanceof TupleDNS.TupleResult, 'Should return TupleResult for range query');
    
    // Cleanup
    for (const coord of coords) {
        await dns.unregister(coord);
    }
});

suite.addTest('Error Handling', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    
    // Test invalid coordinate registration
    try {
        await dns.register('invalid coordinate', ['test']);
        suite.assert(false, 'Should throw error for invalid coordinate');
    } catch (error) {
        suite.assert(true, 'Should throw error for invalid coordinate');
    }
    
    // Test invalid pattern discovery
    try {
        await dns.find('');
        suite.assert(false, 'Should throw error for empty pattern');
    } catch (error) {
        suite.assert(true, 'Should throw error for empty pattern');
    }
});

suite.addTest('Music Collaboration Scenario', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    
    const timestamp = Date.now();
    const musicians = [
        [`ambient.120.${timestamp}.london.music.tuple`, ['midi', 'real-time', 'collaborative']],
        [`jazz.140.${timestamp}.newyork.music.tuple`, ['midi', 'live-recording', 'improvisation']],
        [`electronic.128.${timestamp}.berlin.music.tuple`, ['midi', 'synthesis', 'real-time']]
    ];
    
    // Register musicians
    for (const [coord, caps] of musicians) {
        await dns.register(coord, caps);
    }
    
    // Find MIDI-capable musicians
    const midiResult = await dns.find(`*.*.${timestamp}.*.music.tuple`, ['midi']);
    suite.assert(midiResult instanceof TupleDNS.TupleResult, 'Should find MIDI musicians');
    
    // Find real-time collaborators
    const realtimeResult = await dns.find(`*.*.${timestamp}.*.music.tuple`, ['real-time']);
    suite.assert(realtimeResult instanceof TupleDNS.TupleResult, 'Should find real-time musicians');
    
    // Cleanup
    for (const [coord] of musicians) {
        await dns.unregister(coord);
    }
});

suite.addTest('IoT Device Management Scenario', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    
    const timestamp = Date.now();
    const devices = [
        [`sensor-1.kitchen.floor-1.building-${timestamp}.spatial.tuple`, ['temperature', 'humidity', 'iot']],
        [`printer-1.lab.floor-2.building-${timestamp}.spatial.tuple`, ['3d-print', 'pla', 'petg']],
        [`light-1.office.floor-2.building-${timestamp}.spatial.tuple`, ['lighting', 'dimmer', 'color']]
    ];
    
    // Register devices
    for (const [coord, caps] of devices) {
        await dns.register(coord, caps);
    }
    
    // Find all devices in building
    const buildingResult = await dns.find(`*.*.*.building-${timestamp}.spatial.tuple`);
    suite.assert(buildingResult instanceof TupleDNS.TupleResult, 'Should find building devices');
    
    // Find IoT sensors
    const sensorResult = await dns.find(`*.*.*.building-${timestamp}.spatial.tuple`, ['iot']);
    suite.assert(sensorResult instanceof TupleDNS.TupleResult, 'Should find IoT sensors');
    
    // Cleanup
    for (const [coord] of devices) {
        await dns.unregister(coord);
    }
});

suite.addTest('Performance Test', async () => {
    const dns = new TupleDNS.TupleDNS({
        proxyUrl: `http://localhost:${suite.proxyPort}`
    });
    
    await dns.init();
    
    const startTime = Date.now();
    const coords = [];
    
    // Register 50 nodes
    for (let i = 0; i < 50; i++) {
        const coord = `perf.${i}.${Date.now()}.test.tuple`;
        coords.push(coord);
        await dns.register(coord, ['performance', 'test']);
    }
    
    const regTime = Date.now() - startTime;
    
    const discStart = Date.now();
    
    // Perform 10 discoveries
    for (let i = 0; i < 10; i++) {
        await dns.find('perf.*.*.test.tuple');
    }
    
    const discTime = Date.now() - discStart;
    
    console.log(`  Performance: 50 registrations in ${regTime}ms, 10 discoveries in ${discTime}ms`);
    
    // Cleanup
    for (const coord of coords) {
        await dns.unregister(coord);
    }
    
    suite.assert(true, 'Performance test completed');
});

// Run the test suite
if (require.main === module) {
    suite.runTests().then(success => {
        process.exit(success ? 0 : 1);
    }).catch(error => {
        console.error('Test suite failed:', error);
        process.exit(1);
    });
}

module.exports = TestSuite;